### ISSUE TITLE
Teach the Shake executable to create simple build systems

### ISSUE THREAD

**Issue by snowleopard** [ndmitchell/shake#487]
I wanted the following feature a couple of times: quickly create a simple build system template by running the `shake` executable, so I don't need to do copy-pasting of Shake examples and tweak them. I think other users may want this too, especially if they are new to Shake and want to start using it on a new project. Shake already comes with an executable which could do that. Making this part of Shake also makes sure these templates stay up-to-date.

An example use case:

I'm writing a LaTeX paper. I type something like `shake --new=latex` and Shake goes ahead and generates a working build script that I can hopefully use immediately or tweak it if need be. One could imagine some extra command line options, e.g. `--main=paper.tex` (with default being `main.tex`).

With the above implemented, one could explore other interesting modes, e.g. `shake --add-rule=c-objects` could append a rule for building C object files with dynamic dependencies to an existing build script. In fact, the `--new=latex` could be implemented just as a syntax sugar for assembling a build system out of simple primitives, i.e. `--new=latex` ~ `--add-target=main.pdf --add-rule=pdf-latex`.

This should be easy to implement, so I could even attempt it myself, but there are quite a few design choices to make. Perhaps, this has already been discussed, but I couldn't find an existing issue, although I found #247 slightly relevant as it discusses how to compose build primitives into bigger scripts.

**Comment by ndmitchell** [comment]
My guess is that if the design choices were made cleverly, the template Latex build system would just be:

```
main = buildLatex
```

Perhaps with tweaks, or more options, but I'm guessing any options you put on the command line would correspond 1-to-1 with things that appear in the build system. Once you're at that point, the generator is of marginal use (or if you did want it, you'd want it as a GUI, so probably a web page where you click to generate the build system - so you don't have to learn either the functions or the flags).

But that leads immediately to the question "what should the templates be", and I think that's the really hard thing...

**Comment by snowleopard** [comment]
I though about `main = buildLatex`, but there seem to be some cons compared to generated templates:
* It's not clear how users can customise a standard template without editing Shake's sources. A generated template is really easy to customise -- just edit the file.
* A template also serves an educational purpose for newcomers -- See, that's the whole build system! -- while `buildLatex` is a black box, which may seem magic and dull (nothing to play with). Think `--demo`, but useful for simple everyday build tasks.
* A minor point: `main = buildLatex` also requires typing (i.e. copying) the right `import` statement.

I agree though that the question "what should the templates be" is quite hard. One way to answer this question in a simple way is to postulate: "A template is whatever we accept into /templates directory" :) Note, in the simplest form, a template is generated by just copying a file with a specified name from this directory, leaving all customisation to users.

A webpage for generating build systems sounds interesting too, I haven't thought about this. Can be much more powerful and user-friendly, but is much heavier implementation-wise.

**Comment by ndmitchell** [comment]
Fair enough, I can believe that `buildLatex` is not enough. And it does sound like a template-driven system is the right way forward. I also take the point about education for newcomers - having some "ideal" build systems is valuable. I guess the next step is to write some templates 😃 

It doesn't have to be much heavier implementation-wise to do a web based GUI, provided you have the right abstractions. I even have code to autogenerate a GUI from a web page, although it is quite old and may have rotted (https://github.com/ndmitchell/cmdargs/tree/master/cmdargs-browser).

**Comment by snowleopard** [comment]
Just realised that LaTeX is quite a complicated example :-)

I guess something like this could be a start:
```haskell
import Development.Shake
import System.FilePath

docName :: FilePath
docName = "paper"

main :: IO ()
main = shake shakeOptions { shakeFiles="_build" } $ do
    want [docName <.> "pdf"]

    "*.pdf" %> \pdf -> do
        let src = pdf -<.> "tex"
        need [src]
        () <- cmd "pdflatex" [src]
        () <- cmd "bibtex"   [src]
        cmd "pdflatex" [src]

    "clean" ~> do
        putNormal "Cleaning build products"
        liftIO . removeFiles "." $ map (docName <.>) ["pdf", "aux", "log", "bbl", "blg"]
        putNormal "Cleaning files in _build"
        removeFilesAfter "_build" ["//*"]
```

Ideally, we should also track dynamic dependencies using pdflatex's `-recorder` flag, and perhaps have a rule for rendering images, and maybe ... but at some point this gets a bit too far for a template. This project seems to be trying to do all that: https://github.com/kaoskorobase/shake-latex.

**Comment by ndmitchell** [comment]
So maybe there should be a library for writing specific build systems, e.g. shake-rules, or shake-latex, and then the template stuff should use those?

**Comment by snowleopard** [comment]
@ndmitchell A fully featured LaTeX build system should probably live in a separate package... but then I'm not sure how it will work: there will be a cyclic dependency between `shake` and `shake-latex`?

**Comment by thoughtpolice** [comment]
Well, to be fair, you don't need a direct dependency - `shake --new=latex` could just as easily for example, say, "go to https://github.com/shake-examples/latex-example and download that repository for the template" or something. The template logic doesn't need a literal dependency, just a way to actually copy files into some place, right?

In general, I think it would be nice if there was, along with Shake, a set of 'blessed' corresponding libraries that added basic functionality, like rules and oracles for various projects or tools/languages. Then the minimal 'shake' executable could simply use 'blessed' templates for those projects. Let's say in this theoretical universe that Shake and these libraries were under their own organization, e.g.

```
https://github.com/shakebuild/shake    # the main system
https://github.com/shakebuild/shake-binutils # binutils integration
https://github.com/shakebuild/shake-ccompiler # utilities for using the C compiler
https://github.com/shakebuild/shake-java # basic javac/java integration
https://github.com/shakebuild/shake-latex # rules for latex
```

etc etc. Then you could also have a repository, called like, `github.com/shakebuild/templates`, with a layout like:

```
./latex/
./c/
./java/
./haskell/
./ocaml/
```

Then if I ran `shake --new=ocaml`, it'd just give me a copy of that directory.

Of course, this is just a theoretical idea. I don't endorse moving the repository or anything, but I think in the long run a set of blessed 'extra' libraries are probably a good step forward. I've been thinking of similar things -- I recently wrote a bunch of Shake rules for integration with [Project IceStorm](http://www.clifford.at/icestorm/) and [CLaSH](http://clash-lang.org), and it'd be nice if I could offer them up as a start to building a 'de facto' Shake library for IceStorm, etc.

**Comment by ndmitchell** [comment]
I don't think putting lots of language/target specific stuff in `shake` itself is a good idea, as everyone else seems to agree. I'm not sure if it's better to have `shake-latex` or just `shake-rules` which has the rules for C, Java, Latex etc in them. All the things are going to need a fairly common structure to integrate nicely (even if I don't know what that structure is yet), and the actual rules all being in one project probably isn't too problematic. In that universe, the `shake-rules` package would be the thing which created template projects.
